cleverhans/attacks_tf.py

```diff
@@ -725,9 +757,162 @@ def compare(x, y):
+                                         targeted, learning_rate,
+                                         1, max_iterations,
+                                         abort_early, initial_const,
+                                         clip_min, clip_max, num_labels, shape,
+                                         extension=0)
+
+    def attack(self, instances, targets):
+        """
+        Perform the L_0 attack on the given instance for the given targets.
+
+        If self.targeted is true, then the targets represents the target labels
+        If self.targeted is false, then targets are the original class labels
+        """
+
+        r = []
+        for i, (instance, target) in enumerate(zip(instances, targets)):
+            _logger.debug(("Running CWL0 attack on instance " +
+                           "{} of {}").format(i, len(instances)))
+            r.extend(self.attack_single(instance, target))
+        return np.array(r)
+
+    def attack_single(self, instance, target):
+        """
+        Run the attack on a single instance and label.
+        """
+
+        # the pixels we can change
+        valid = np.ones(instance.shape)
+
+        # the previous image
+        prev = np.copy([instance])
+
+        last_solution = [instance]
+        const = self.initial_const
+        equal_count = 0
+
+        while True:
+            # try to solve given this valid map
+            self.l2_attack.source_image = np.copy(prev)
+            while const < self.largest_const:
+                # try solving for each value of the constant
+                self.l2_attack.initial_const = const
+                res = self.l2_attack.attack_batch(np.copy([instance]),
+                                                  np.array([target]),
+                                                  mask=np.array([valid]))
+                if res is not None:
+                    break
+                const *= self.const_factor
+
+            if res is None:
+                # the attack failed, we return this as our final answer
+                _logger.debug("Attack succeeded with {} fixed values.".
+                              format(equal_count))
+                return last_solution
+
+            # the attack succeeded, now we pick new pixels to set to 0
+            restarted = False
+            gradientnorm, scores, nimg = res
+
+            equal_count = np.sum(np.abs(instance-nimg[0]) < .0001)
+            if np.sum(valid) == 0:
+                # if no pixels changed, return
+                return [instance]
+            _logger.debug("Next iteration; {} fixed values."
+                          .format(equal_count))
+
+            orig_shape = valid.shape
+            valid = valid.flatten()
+            totalchange = abs(nimg[0]-instance)*np.abs(gradientnorm[0])
+            totalchange = totalchange.flatten()
+
+            # set some of the pixels to 0 depending on their total change
+            did = 0
```

Comment:
Is there a more descriptive name for this variable other than did? I'm not sure what this is tracking by just looking at the code.